Enable preferred-greeters with lightdm

adriaandegroot opened the pull request:
This builds on @paolodongilli PR and makes things configurable, while improving the code that is looking for the greeter a little bit. Since we have support for per-DM configuration now, it's relatively simple to add this to the configuration file.

paolodongilli committed:
Add detection of default lightdm-greeter

adriaandegroot committed:
[displaymanager] Factor out the find-a-greeter part

adriaandegroot committed:
[displaymanager] Handle preferred / deterministic greeter

adriaandegroot committed:
[displaymanager] Allow configuring the preferred-greeter-list

adriaandegroot commented:
This PR takes over #2040

dalto8 commented on src/modules/displaymanager/main.py:
```diff
@@ -544,6 +544,9 @@ class DMlightdm(DisplayManager):
     executable = "lightdm"
 
+    # Can be overridden in the .conf file
+    preferred_greeters = ["lightdm-greeter.desktop"]
```
Since `lightdm-greeter` is debian specific, maybe `lightdm-gtk-greeter` is a more reasonable default?

adriaandegroot replied:
I've picked "empty list" as default so it doesn't have **any** preferred one, which falls back to the older behavior (except now it's deterministic by alphabet).

dalto8 commented on src/modules/displaymanager/main.py:
```diff
@@ -633,41 +636,57 @@ def desktop_environment_setup(self, default_desktop_environment):
+        desktop_names = [n for n in available_names if n.endswith(".desktop")] # .. otherwise any .desktop
+        if desktop_names:
+            return desktop_names[0]
+        desktop_names = [n for n in available_names if not n.startswith(".")] # .. otherwise any non-dot-file
+        if desktop_names:
+            return desktop_names[0]
+        if available_names: # Anything?
+            return available_names[0]
```
Are these conditions valid?  Should we just be throwing an exception or returning an empty array here?

adriaandegroot replied:
Good point, the last one doesn't make any sense (the one on L660 does, if none of the preferred ones is matched).

dalto8 replied:
Isn't the one at line 659 also invalid?  That looks like it accepts any file that doesn't start with `.`

dalto8 commented on src/modules/displaymanager/main.py:
```diff
@@ -633,41 +636,57 @@ def desktop_environment_setup(self, default_desktop_environment):
+                greeter = greeter[:-8] # Remove ".desktop" from end
+
+            libcalamares.utils.debug("found greeter {!s}".format(greeter))
+            os.system(
+                "sed -i -e \"s/^.*greeter-session=.*"
+                "/greeter-session={!s}/\" {!s}".format(
+                    greeter,
+                    lightdm_conf_path
+                )
```
This should probably be done in python instead of using `sed` but at the very least it should be calling a libcalamares function instead of `os.system`.

adriaandegroot replied:
We don't really have a sed-like thing, do we (maybe we can link to a libsed or some suitable line-by-line replacer?). This module and the tracking module (! in C++) are the primary users of sed, so that would be a good improvement (also, eventually, towards dropping root).

But I'm not going to fix it in this PR, that's a larger issue.

adriaandegroot committed:
[displaymanager] lightdm: default to no-preferred-greeter

While here, drop nonsensical alternatives: if it's not a
.desktop file, it's not going to work anyway.

adriaandegroot committed:
[displaymanager] lightdm: logging errors

Don't just use *any* file found in the xgreeters directory,
and log when no greeter was found (outside of the translated
error message).
